contributors/design-proposals/storage/container-storage-interface-pod-information.md

```diff
@@ -0,0 +1,123 @@
+        // Indicates if kubelet shall send Pod in NodePublish requests
+        bool requires_publish_pod = 1;
+}
+
+
+// CSIDriverInfo is the empty request message from Kubelet
+message CSIDriverInfoRequest {
+}
+
+// Registration is the service advertised by the Plugins.
+service CSIRegistration {
+        rpc GetCSIDriverInfo(CSIDriverInfoRequest) returns (CSIDriverInfo) {}
+}
+```
+
+This API will be accompanied with `hack/update-generated-csi-registration.sh` and `hack/update-generated-csi-registration-dockerized.sh` to generate golang code for the API definition (based on `hack/update-generated-device-plugin*.sh`).
+
+
+#### CSI volume plugin
+* It already has a map of registered CSI drivers and their sockets. This map must be extended with a flag if a driver wants to receive pod information in `NodePublish`.
+* It implements client side of `CSIRegistration` service. When the plugin gets a callback that a new CSI socket (created by Driver Registrar) was detected, it uses this socket to send `CSIDriverInfoRequest` to the Driver Registrar. The Driver Registrar either returns `CSIDriverInfo` with the flag or the Driver Registrar does not implement the method and we can assume that the corresponding CSI driver does not want to receive pod information.
+* Based on the flag above, it passes pod information in `NodePublish` requests.
+
+#### CSI Driver Registrar
+It runs as a sidecar container in every pod with a CSI driver on every node. Each CSI driver runs its own CSI Driver Registrar.
+* It has a new command line option `--requires-publish-pod` that admin (or CSI driver author) sets if the CSI driver needs pod information in `NodePublish`.
+* It implements server part of `CSIRegistration` service, i.e. listens for  `GetCSIDriverInfo` call and returns `CSIDriverInfo` with the flag.
+
+
+## End to end examples
+
+### CSI driver wants pod information in `NodePublish`
+1. CSI driver author provides documentation and templates that the driver need  pod information in `NodePublish`.
+2. Admin installs CSI driver with CSI Driver Registrar `--requires-publish-pod`, e.g. by applying helm template from CSI diver author.
+3. On a node, CSI Driver Registrar places UNIX domain socket to `/var/lib/kubelet/plugins/foo-reg.sock`.
+4. Kubelet finds the socket there and calls `GetInfo` on it (no change from current behavior).
+5. CSI Driver Registrar receives `GetInfo` call and passes CSI driver socket to kubelet (no change from current behavior).
+6. Kubelet notifies CSI volume plugin that there is a new CSI driver installed and passes **both** CSI DriverRegistrar socket and CSI driver socket (no change from current behavior).
+7. CSI volume plugin calls `GetCSIDriverInfo` on the CSI Driver Registrar socket.
+8. CSI Driver Registrar, based on `--requires-publish-pod` presence on command line, returns `CSIDriverInfo` with `requires_publish_pod = true`.
+9. CSI volume plugin stores this flag for the CSI driver.
+
+When a volume is being mounted for a pod:
+
```

Comment:
optional csi stage call here

Reply:
I added NodeStageVolume to the example, with a note that it won't get any pod information.